traitcuration/traits/datasources/zooma.py

```diff
@@ -0,0 +1,84 @@
+"""
+This module contains all the functionality that uses ZOOMA to retrieve mapping suggestions and create the suggested
+terms in the app's database
+"""
+import requests
+
+from ..models import Trait, MappingSuggestion, OntologyTerm, User
+from .ols import make_ols_query, get_ontology_id
+
+
+BASE_URL = "http://www.ebi.ac.uk/spot/zooma/v2/api"
+
+
+def get_zooma_suggestions():
+    """
+    Retrieves zooma mapping suggestions for all traits and creates terms and suggestions in the app's database for each
+    of them.
+    """
+    traits = Trait.objects.all()
+    for trait in traits:
+        results_list = get_suggestions_for_trait(trait)
+        for result in results_list:
+            term = create_local_term(result)
+            # If local term creation returns None, it means that this term already exists in the database
+            if term is not None:
+                create_mapping_suggestion(trait, term)
+
+
+def get_suggestions_for_trait(trait):
+    """
+    Takes in a trait object as its argument and returns the zooma response of a suggestion list as a dictionary.
+    """
+    formatted_trait_name = trait.name.replace(' ', '+')
+    response = requests.get(
+        f"{BASE_URL}/services/annotate?propertyValue={formatted_trait_name} \
+        &filter=required:[none],ontologies:[efo,mondo,hp,orphanet]")
+    return response.json()
+
+
+def create_local_term(result):
+    """
+    Takes in a single zooma suggestion result and creates an ontology term for it in the app's database.
+    """
+    term_iri = result["semanticTags"][0]
+    print(term_iri)
+    if OntologyTerm.objects.filter(iri=term_iri).exists():
+        return
+    # Search for the term in EFO and return its information. If it is not in EFO, search for it in original ontology.
+    term_info = make_ols_query(term_iri, 'efo')
+    term_ontology_id = 'efo'
+    # None here means that the term wasn't found in the EFO ontology
+    if term_info is None:
+        term_ontology_id = get_ontology_id(term_iri)
+        term_info = make_ols_query(term_iri, term_ontology_id)
+        # If the term is not found in the original ontology either, return
+        if term_info is None:
+            print(f'No info found on {term_iri}')
+            return
+    term_status = get_term_status(term_ontology_id, term_info['is_obsolete'])
+    # Create an ontology term in the database
+    term = OntologyTerm(curie=term_info['curie'], iri=term_iri, label=term_info['label'], status=term_status)
+    term.save()
+    return term
+
+
+def get_term_status(ontology_id, is_obsolete):
+    """
+    Takes the ontology_id of a term and a whether it is obsolete or not, and returns its calculated status.
+    'Obsolete' if the is_obsolete flag is true, 'Current' if its ontology is EFO, and 'Needs Import' otherwise
+    """
+    if is_obsolete:
+        return 'obsolete'
```

Comment:
This should ideally use not the literal "obsolete", but some enum value which is defined just once for the entire project. I think we already have one somewhere in data models

Reply:
OK so the enum situation is kind of confusing right now. This is what we currently have:

- A list of possible status fields (values and human-readable labels) for traits and terms *outside* of the model classes, so that it can be shared between those two, and just have the trait status field add another value to this array, for the 'Awaiting Review' status.
```
STATUS_FIELDS = [
    ('current', 'Current'),
    ('unmapped', 'Unmapped'),
    ('obsolete', 'Obsolete'),
    ('deleted', 'Deleted'),
    ('needs_import', 'Needs Import'),
    ('awaiting_import', 'Awaiting Import'),
    ('needs_creation', 'Needs Creation'),
    ('awaiting_creation', 'Awaiting Creation'),
]
```

- Models using these choices as such:
 ```
class Trait(models.Model):
    status = models.CharField(max_length=50, choices=STATUS_FIELDS + [('awaiting_review', 'Awaiting Review')])
    
class OntologyTerm(models.Model):
    status = models.CharField(max_length=50, choices=STATUS_FIELDS)
```

This provides enum functionality and validation if a value which is not in STATUS_FIELDS is provided. 

The problem with that if we don't want to use literals to insert values, especially since the STATUS_FIELDS array is defined outside of each class, is that getting the enum value defined in the models is kind of hard to read and hard to use. For example, to get all possible values one would have to do `Trait._meta.get_field('status').choices` which would return the `STATUS_FIELDS` array exactly as defined, and then use array and tuple methods to get the value.

A possible alternative to that, as described here https://adamj.eu/tech/2020/01/27/moving-to-django-3-field-choices-enumeration-types/ is to create a TextChoices subclass for the status values:
```
from django.db import models


class Status(models.TextChoices):
    CURRENT = 'C', 'Current'
    UNMAPPED = 'U', 'Unmapped'


class Trait(models.Model):
    status = models.CharField(
        max_length=2,
        choices=Status.choices,
        default=Status.UNMAPPED,
    )
```

It can then be accessed like this `unmapped_traits = Trait.objects.filter(status=Status.UNMAPPED)`
The problem with that is that I'm not sure how we would add the 'awaiting_review' specifically to the Trait model, maybe another class that inherits?

Reply:
Yes, sorry, I didn't mean to imply that we would get the values _from_ the models, understandably this could be cumbersome. What I meant is that the `'current'` literal should only be defined at one place in the entire code, and then used by import from there.

The simplest solution I had in mind for that is converting the `STATUS_FIELDS` tuple into a plain [enum](https://docs.python.org/3/library/enum.html) and then, rather than using `'current'` inside zooma.py, use something like `STATUS_FIELDS.CURRENT.value`

The models.TextChoices class which you found looks very similar to what plain enums do. You can choose to use either it or enum, both seem fine to me.

Finally, regarding appending "awaiting review", I think a class which inherits from the main one is a good idea. I think enums support inheritance, and the TextChoices should certainly do since this is just a class

Reply:
https://docs.python.org/3/library/enum.html#restricted-enum-subclassing Turns out ENUM subclassing is actually not possible

Reply:
> Turns out enum field inheritance is actually not possible
Huh. Indeed. Didn't know that.

I'm really not happy with duplicating the core status values. There are two ways we can handle this:
* The author of the enum library suggests going with another library in such cases, aenum (advanced enumerations), which allows extensions and combinations. See his comments here: https://stackoverflow.com/a/33680929, https://stackoverflow.com/a/46080827.
* Alternatively, we could just have a single enum with statuses (including "awaiting_review"), and when defining a model, we could compute the list of allowable values based on that enum, taking all values _except_ the ones we don't want to be allowed.

Reply:
So the approach I've decided to use is mostly the second one you mentioned. I am defining a separate status class with all the possible statuses, and then two class methods, `trait_choices()` and `term_choices()` that filter the possible status values accordingly, so then the code in the Trait and Term models looks like this: 
`status = models.CharField(max_length=50, choices=Status.trait_choices())`

Reply:
This sounds good! We would still be able to access individual statuses as well via some attributes, like `Status.CURRENT`, right? Since that was the whole point

Reply:
Yes yes, we will just need to import the Status class from `models.py` to do that